manager/integration/framework/node_exec/node_exec.py

```diff
@@ -0,0 +1,116 @@
+from kubernetes import client
+from kubernetes.stream import stream
+
+import time
+
+DEFAULT_POD_TIMEOUT = 180
+DEFAULT_POD_INTERVAL = 1
+
+class NodeExec:
+
+    _instance = None
+    node_exec_pod = {}
+
+    def __new__(cls, *args, **kwargs):
+        if cls._instance is None:
+            cls._instance = super().__new__(cls)
+        return cls._instance
+
+    def __init__(cls):
+        cls.core_api = client.CoreV1Api()
+
+    def cleanup(cls):
+        for pod in cls.node_exec_pod.values():
+            res = cls.core_api.delete_namespaced_pod(
+                name=pod.metadata.name,
+                namespace='default',
+                body=client.V1DeleteOptions()
+            )
+            print(res)
+
+    def issue_cmd(cls, node_name, cmd):
+        pod = cls.launch_pod(node_name)
+        exec_command = [
+            'nsenter',
+            '--mount=/rootfs/proc/1/ns/mnt',
+            '--',
+            'sh',
+            '-c',
+            cmd
+        ]
+        res = stream(
+            cls.core_api.connect_get_namespaced_pod_exec,
+            pod.metadata.name,
+            'default',
+            command=exec_command,
+            stderr=True,
+            stdin=False,
+            stdout=True,
+            tty=False
+        )
+        return res
+
+    def launch_pod(cls, node_name):
```

Comment:
It seems you would like to reuse a singleton pod to run commands instead of starting several pods. For the parallel execution in the future, please do concurrent handle for this method like mutex.

Reply:
I make it namespaced and every test case holds it own node_exec, volume, etc class instances to make it possible to run parallel, and these classes don't need to be singleton anymore.